File: src/main/java/com/rs/game/content/world/areas/varrock/npcs/UristLoric.java
package com.rs.game.content.world.areas.varrock.npcs;

import com.rs.engine.dialogue.Dialogue;
import com.rs.engine.dialogue.HeadE;
import com.rs.engine.miniquest.Miniquest;
import com.rs.engine.quest.Quest;
import com.rs.game.World;
import com.rs.game.model.entity.npc.NPC;
import com.rs.game.model.entity.player.Player;
import com.rs.game.model.object.GameObject;
import com.rs.game.tasks.WorldTasks;
import com.rs.lib.game.GroundItem;
import com.rs.lib.game.Tile;
import com.rs.plugin.annotations.PluginEventHandler;
import com.rs.plugin.handlers.NPCClickHandler;
import com.rs.utils.Ticks;

import java.util.Objects;

@PluginEventHandler
public class UristLoric {
    private static final int npcid = 11270;

    public static NPCClickHandler UristLoric = new NPCClickHandler(new Object[]{ npcid }, new String[]{"Talk-to"}, e -> {
        if (!e.getPlayer().isMiniquestStarted(Miniquest.FROM_TINY_ACORNS)) {
            //preQuest(e.getPlayer());
            return;
        }
        switch (e.getPlayer().getQuestStage(Quest.BUYERS_AND_CELLARS)) {
            case 1 -> stage1(e.getPlayer(), e.getNPC());
            //case 2 -> stage2(e.getPlayer());
            //default -> stage3(e.getPlayer());
        }
    });

    public static Tile getFloorDrop(Player player) {
        for (GroundItem groundItem : World.getAllGroundItemsInChunkRange(823724, 1)) {
            if (groundItem == null || groundItem.getDefinitions() == null || groundItem.getId() != 18649) {
                continue;
            }
            if (!groundItem.getTile().withinArea(3220, 3427, 3228, 3432)) {
                player.sendMessage("I should find a suitable spot to put this. Maybe just north of him…");
                continue;
            }
            return Tile.of(groundItem.getTile());
        }
        return null;
    }

    public static void stage1(Player player, NPC npc) {
        if(getFloorDrop(player) != null){
            player.startConversation(new Dialogue()
                    .addPlayer(HeadE.CALM_TALK, "That thing on the ground… Is it yours?")
                    .addNPC(npcid, HeadE.SCARED, "What thing? Oh! Thank you kindly, I'd have been sad if I'd lost that.", () -> {
                        npc.walkToAndExecute(getFloorDrop(player), () ->{
                            int loop = 0;
                            WorldTasks.schedule(Ticks.fromSeconds(5),Ticks.fromSeconds(5), () -> {
                            switch (loop) {
                                case 0 -> {
                                    npc.forceTalk("Hmm, how'd this get over here?");
                                    //World.removeGroundItem(player, item at tile getFloorDrop(player))
                                }
                            }
                            });
                        });
                    }));
            return;
        }
        if(player.getInventory().containsItem(18651)) {
            player.startConversation(new Dialogue()
                    .addPlayer(HeadE.CALM_TALK, "I'm just checking up on progress for Mr. Lightfinger. You said it would be ready in a couple of days?")
                    .addNPC(npcid, HeadE.HAPPY_TALKING, "Aye, it's nearly done. It's right he-")
                    .addPlayer(HeadE.CONFUSED, "...")
                    .addNPC(npcid, HeadE.SCARED, "By all the gods! Where's it gone? Where's it got to?")
                    .addPlayer(HeadE.CONFUSED, "It's not walked off, has it?")
                    .addNPC(npcid, HeadE.FRUSTRATED, "It couldn't walk that far on one turn of the spring.")
                    .addPlayer(HeadE.CALM_TALK, "I hope it turns up, then; it's due very soon. Could it have been stolen?")
                    .addNPC(npcid, HeadE.FRUSTRATED, "In Varrock? With that guard watching like a hawk?")
                    .addPlayer(HeadE.CALM_TALK, "He doesn't seem all that hawklike.")
                    .addNPC(npcid, HeadE.SKEPTICAL, "Now that you mention it…")
                    .addPlayer(HeadE.CALM_TALK, "Will you be able to finish by the deadline?")
                    .addNPC(npcid, HeadE.SHAKING_HEAD, "Are you serious? Not a chance!")
                    .addPlayer(HeadE.CALM_TALK, "Then what will you tell Mr. Lightfinger?")
                    .addNPC(npcid, HeadE.VERY_FRUSTRATED, "I don't see that I have much choice; I shall have to give him his money back.")
                    .addPlayer(HeadE.CALM_TALK, "That sounds fair.")
                    .addNPC(npcid, HeadE.HAPPY_TALKING, "Since you're working for him, can you take this banker's note to him with my sincerest apologies?")
                    .addPlayer(HeadE.CALM_TALK, "Urist hands you a banker's note. The figure on it is astronomical.", () -> {
                        player.getInventory().addItem(18652, 1); // Add Banker's note to player's inventory
                        player.getMiniquestManager().setStage(Miniquest.FROM_TINY_ACORNS, 2);
                    })
                    .addPlayer(HeadE.CALM_TALK, "I should think so.")
                    .addNPC(npcid, HeadE.CALM_TALK, "I appreciate your understanding.")
                    .addPlayer(HeadE.CALM_TALK, "I appreciate your cooperation.")
            );
        }
        else {
            player.startConversation(new Dialogue()
                    .addPlayer(HeadE.CALM_TALK, "Are you the master craftsman working on a commission for Darren Lightfinger?")
                    .addNPC(npcid, HeadE.HAPPY_TALKING, "That I am. What can I do for you?")
                    .addPlayer(HeadE.CALM_TALK, "How's it coming along?")
                    .addNPC(npcid, HeadE.HAPPY_TALKING, "It's very nearly done. Just needs a bath in preserving oil to protect the mechanisms, and then a good polish.")
                    .addPlayer(HeadE.CALM_TALK, "Is this it here on your stall?")
                    .addNPC(npcid, HeadE.HAPPY_TALKING, "Aye, the baby red dragon there. As ordered, its scales are perfect rubies and it's capable of walking and breathing fire.")
                    .addNPC(npcid, HeadE.HAPPY_TALKING, "Could have made it fly too, given another six months or so.")
                    .addPlayer(HeadE.CALM_TALK, "Very nice. Well, see you later.")
                    .addNPC(npcid, HeadE.HAPPY_TALKING, "Aye, that you will.")
            );
        }
    }
}

File: src/main/java/com/rs/game/content/world/areas/varrock/npcs/UristLoric.kt
package com.rs.game.content.world.areas.varrock.npcs

import com.rs.engine.dialogue.HeadE
import com.rs.engine.dialogue.startConversation
import com.rs.engine.miniquest.Miniquest
import com.rs.game.World
import com.rs.game.content.miniquests.from_tiny_acorns.PickPocketUrist
import com.rs.game.model.entity.async.schedule
import com.rs.game.model.entity.npc.NPC
import com.rs.game.model.entity.player.Player
import com.rs.lib.game.GroundItem
import com.rs.lib.game.Tile
import com.rs.plugin.annotations.ServerStartupEvent
import com.rs.plugin.kts.onNpcClick
import com.rs.utils.Ticks

private const val talisman = 18649
private const val bankNote = 18652
private const val babyDragon = 18651
private const val waitTime = 5

private val uristTile: Tile = Tile.of(3222, 3424, 0)

@ServerStartupEvent
fun mapUristPlugins() {
    onNpcClick(11270, options = arrayOf("Talk-to")) { (player, npc) ->
        if (npc.tile != uristTile) {
            player.sendMessage("He looks busy.")
            return@onNpcClick
        }
        if (!player.isMiniquestStarted(Miniquest.FROM_TINY_ACORNS)) {
            preQuest(player)
            return@onNpcClick
        }
        if (player.isMiniquestComplete(Miniquest.FROM_TINY_ACORNS)) {
            postQuest(player, npc)
            return@onNpcClick
        }
        when (player.getMiniquestStage(Miniquest.FROM_TINY_ACORNS)) {
            1 -> stage1(player, npc)
            2 -> stage2(player, npc)
            3 -> stage3(player, npc)
            else -> postQuest(player, npc)
        }
    }

    onNpcClick(11270, options = arrayOf("Pickpocket")) { (p, npc) ->
        if (!p.isMiniquestStarted(Miniquest.FROM_TINY_ACORNS)) {
            p.startConversation {
                npc(npc, HeadE.ANGRY, "Oi! Leave that alone.")
            }
            return@onNpcClick
        }
        if (p.inventory.containsItem(talisman)) {
            p.sendMessage("You've stolen his talisman already.")
            return@onNpcClick
        }
        if (getFloorDrop(p) != null) {
            p.sendMessage("His talisman isn't in his pocket, it's on the ground nearby.")
            return@onNpcClick
        }
        if (p.inventory.containsItem(babyDragon) || p.miniquestManager.isComplete(Miniquest.FROM_TINY_ACORNS)) {
            p.sendMessage("You don't need this; you've already got the dragon.")
            return@onNpcClick
        }
        p.actionManager.setAction(PickPocketUrist(npc))
    }
}

private fun getFloorDrop(player: Player): GroundItem? {
    for (groundItem in World.getAllGroundItemsInChunkRange(823724, 1)) {
        if (groundItem.definitions == null || groundItem.id != talisman) {
            continue
        }
        if (!groundItem.tile.withinArea(3220, 3427, 3228, 3432)) {
            player.sendMessage("I should find a suitable spot to put this. Maybe just north of him...")
            continue
        }
        return groundItem
    }
    return null
}

private fun preQuest(player: Player) {
    player.sendMessage("The Dwarf is hard at work on some sort of highly delicate construction, and doesn't pay you any notice.")
}

private fun stage1(player: Player, npc: NPC) {
    getFloorDrop(player)?.let { floorDrop ->
        player.startConversation {
            player(HeadE.CALM_TALK, "That thing on the ground... Is it yours?")
            npc(npc, HeadE.SCARED, "What thing? Oh! Thank you kindly, I'd have been sad if I'd lost that.")
            exec {
                npc.walkToAndExecute(floorDrop.tile) {
                    player.schedule {
                        for (i in 0..3) {
                            when (i) {
                                0 -> {
                                    player.miniquestManager.getAttribs(Miniquest.FROM_TINY_ACORNS).setB("UristDistracted", true)
                                    npc.forceTalk("Hmm, how'd this get over here?")
                                    World.removeGroundItem(floorDrop)
                                    wait(Ticks.fromSeconds(waitTime))
                                }
                                1 -> {
                                    npc.forceTalk("Ugh it's all dirty. Lucky I've got my blue silk handkerchief on me.")
                                    wait(Ticks.fromSeconds(waitTime))
                                }
                                2 -> {
                                    npc.forceTalk("There, that's better.")
                                    wait(Ticks.fromSeconds(waitTime))
                                }
                                3 -> {
                                    npc.forceTalk("Well, back to work.")
                                    player.miniquestManager.getAttribs(Miniquest.FROM_TINY_ACORNS).setB("UristDistracted", false)
                                    wait(Ticks.fromSeconds(waitTime))
                                }
                            }
                        }
                        npc.walkToAndExecute(uristTile) { npc.resetDirection() }
                        player.unlock()
                    }
                }
            }
        }
    } ?: run {
        // No floor drop present, proceed with normal conversation
        player.startConversation {
            player(HeadE.CALM_TALK, "Are you the master craftsman working on a commission for Darren Lightfinger?")
            npc(npc, HeadE.HAPPY_TALKING, "That I am. What can I do for you?")
            player(HeadE.CALM_TALK, "How's it coming along?")
            npc(npc, HeadE.HAPPY_TALKING, "It's very nearly done. Just needs a bath in preserving oil to protect the mechanisms, and then a good polish.")
            player(HeadE.CALM_TALK, "Is this it here on your stall?")
            npc(npc, HeadE.HAPPY_TALKING, "Aye, the baby red dragon there. As ordered, its scales are perfect rubies and it's capable of walking and breathing fire.")
            npc(npc, HeadE.HAPPY_TALKING, "Could have made it fly too, given another six months or so.")
            player(HeadE.CALM_TALK, "Very nice. Well, see you later.")
            npc(npc, HeadE.HAPPY_TALKING, "Aye, that you will.")
        }
    }
}



private fun stage2(player: Player, npc: NPC) {
    if (player.inventory.containsItem(babyDragon) || player.bank.containsItem(babyDragon, 1)) {
        player.startConversation {
            player(HeadE.CALM_TALK, "I'm just checking up on progress for Mr. Lightfinger. You said it would be ready in a couple of days?")
            npc(npc, HeadE.HAPPY_TALKING, "Aye, it's nearly done. It's right he-")
            player(HeadE.CONFUSED, "...")
            npc(npc, HeadE.SCARED, "By all the gods! Where's it gone? Where's it got to?")
            player(HeadE.CONFUSED, "It's not walked off, has it?")
            npc(npc, HeadE.FRUSTRATED, "It couldn't walk that far on one turn of the spring.")
            player(HeadE.CALM_TALK, "I hope it turns up, then; it's due very soon. Could it have been stolen?")
            npc(npc, HeadE.FRUSTRATED, "In Varrock? With that guard watching like a hawk?")
            player(HeadE.CALM_TALK, "He doesn't seem all that hawklike.")
            npc(npc, HeadE.SKEPTICAL, "Now that you mention it...")
            player(HeadE.CALM_TALK, "Will you be able to finish by the deadline?")
            npc(npc, HeadE.SHAKING_HEAD, "Are you serious? Not a chance!")
            player(HeadE.CALM_TALK, "Then what will you tell Mr. Lightfinger?")
            npc(npc, HeadE.VERY_FRUSTRATED, "I don't see that I have much choice; I shall have to give him his money back.")
            player(HeadE.CALM_TALK, "That sounds fair.")
            npc(npc, HeadE.HAPPY_TALKING, "Since you're working for him, can you take this banker's note to him with my sincerest apologies?")
            simple("Urist hands you a banker's note. The figure on it is astronomical.") {
                player.inventory.addItem(18652, 1)
                player.miniquestManager.setStage(Miniquest.FROM_TINY_ACORNS, 3)
            }
            player(HeadE.CALM_TALK, "I should think so.")
            npc(npc, HeadE.CALM_TALK, "I appreciate your understanding.")
            player(HeadE.CALM_TALK, "I appreciate your cooperation.")
        }
    } else {
        player.startConversation {
            player(HeadE.CALM_TALK, "Oh, nice dragon.")
            npc(npc, HeadE.HAPPY_TALKING, "I thought I'd lost it for a while. Lucky I found it, eh?") {
                player.miniquestManager.setStage(Miniquest.FROM_TINY_ACORNS, 1)
            }
        }
    }
}

private fun stage3(player: Player, npc: NPC) {
    player.startConversation{
        if (!player.inventory.containsItem(babyDragon) && !player.bank.containsItem(babyDragon, 1)) {
                player(HeadE.CALM_TALK, "Oh, nice dragon.")
                npc(npc, HeadE.HAPPY_TALKING, "I thought I'd lost it for a while. Lucky I found it, eh?") {
                    player.miniquestManager.setStage(Miniquest.FROM_TINY_ACORNS, 1)
                    if (player.inventory.containsItem(bankNote)) {
                        player.inventory.deleteItem(bankNote, 1)
                        player.sendMessage("The banker's note is now useless to you; you crumple it and throw it away.")
                    }
                    if (player.bank.containsItem(bankNote, 1)) {
                        player.bank.deleteItem(bankNote, 1)
                        player.sendMessage("The banker's note is now useless, I'm sure the bank will dispose of it.")
                    }
                }
                player(HeadE.FRUSTRATED, "How... fortunate.")
            return@startConversation
        }
        if (!player.inventory.containsItem(bankNote) && !player.bank.containsItem(bankNote, 1)) {
            player(HeadE.CALM_TALK, "I'm afraid I can't find the banker's note you made out.")
            npc(npc, HeadE.FRUSTRATED, "Then I'll have to write another.") {
                player.inventory.addItem(bankNote)
                player.sendMessage("Urist hands you another banker's note. The figure on it is still astronomical.")
            }
            return@startConversation
        }
        npc(npc, HeadE.SAD_CRYING, "Ruined! Ruined! It makes me want to throw a tantrum!")
    }
}

fun postQuest(player: Player, npc: NPC) {
    player.startConversation {
        npc(npc, HeadE.SAD_CRYING, "Ruined! Ruined! It makes me want to throw a tantrum!")
    }
}
